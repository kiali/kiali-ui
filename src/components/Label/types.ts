export interface Labels {
  name: string;
  value: string;
}
