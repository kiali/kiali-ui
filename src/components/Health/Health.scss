svg:not(:root).icon-failure,
svg:not(:root).icon-degraded,
svg:not(:root).icon-healthy,
svg:not(:root).icon-na {
  overflow: visible;
}
.tippy-popper[x-placement^='left'] .pf-tooltip-theme .tippy-arrow {
  background-color: #fff;
  border-color: black;
  border-width: 1px;
  border-style: solid;
}
.health_indicator .pf-c-tooltip__content {
  background-color: #fff;
  border-style: solid;
  border-color: black;
  border-width: 1px;
  text-align: left;
}
