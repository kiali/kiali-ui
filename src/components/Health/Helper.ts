import * as React from 'react';
import { Status } from 'types/Health';

type Size = 'sm' | 'md' | 'lg' | 'xl';

export const createIcon = (status: Status, size?: Size) => {
  return React.createElement(status.icon, { color: status.color, size: size, className: status.class });
};
