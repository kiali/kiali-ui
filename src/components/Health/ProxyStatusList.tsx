import * as React from 'react';
import { ProxyStatus } from '../../types/Health';
import { Flex, FlexItem, Split, SplitItem, Text, TextVariants } from '@patternfly/react-core';
import { style } from 'typestyle';

type Props = {
  workloadName: string;
  statuses: ProxyStatus[];
};

const smallStyle = style({ fontSize: '70%' });
const workloadNameStyle = style({ minWidth: '6rem', overflowWrap: 'break-word' });

const proxyStatusSort = (a: ProxyStatus, b: ProxyStatus): number => {
  return a.component > b.component ? 1 : -1;
};

class ProxyStatusList extends React.Component<Props> {
  sortedStatus = () => {
    return this.props.statuses.sort(proxyStatusSort);
  };

  statusList = () => {
    const statusItems = this.sortedStatus().map((value: ProxyStatus, i: number) => {
      return (
        <FlexItem key={'proxy-status-' + i}>
          <Text component={TextVariants.small} className={smallStyle}>
            {value.component + ': ' + value.status}
          </Text>
        </FlexItem>
      );
    });

    return <Flex style={{ flex: 'inlineFlex' }}> {statusItems} </Flex>;
  };

  render() {
    if (this.props.statuses.length > 0) {
      return (
        <Split gutter={'sm'}>
          <SplitItem className={workloadNameStyle}>{this.props.workloadName}</SplitItem>
          <SplitItem>{this.statusList()}</SplitItem>
        </Split>
      );
    } else {
      return [];
    }
  }
}

export default ProxyStatusList;
