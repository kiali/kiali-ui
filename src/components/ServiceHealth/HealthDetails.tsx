import * as React from 'react';
import { Icon, OverlayTrigger, Popover } from 'patternfly-react';

import { Health } from '../../types/Health';
import { getName } from '../../types/RateIntervals';
import * as H from '../../utils/Health';

interface Props {
  id: string;
  health: Health;
  headline: string;
  placement?: string;
  rateInterval: number;
}

export class HealthDetails extends React.PureComponent<Props, {}> {
  constructor(props: Props) {
    super(props);
  }

  render() {
    return (
      <OverlayTrigger
        placement={this.props.placement || 'right'}
        overlay={this.tooltipContent()}
        trigger={['hover', 'focus']}
        rootClose={false}
      >
        {this.props.children}
      </OverlayTrigger>
    );
  }

  tooltipContent() {
    const health = this.props.health;
    const deplsStatuses = health.deploymentStatuses.map(st => H.ratioCheck(st.available, st.replicas));
    const allInactive = deplsStatuses.length > 0 && !deplsStatuses.some(st => st !== H.NA);
    const globalDeplStatus: H.Status = allInactive
      ? H.FAILURE
      : deplsStatuses.reduce((prev, cur) => H.mergeStatus(prev, cur), H.NA);
    const envoyInbound = H.ratioCheck(health.envoy.inbound.healthy, health.envoy.inbound.total);
    const envoyOutbound = H.ratioCheck(health.envoy.outbound.healthy, health.envoy.outbound.total);
    const reqErrorsRatio = H.getRequestErrorsRatio(health.requests);
    const reqErrorsText = reqErrorsRatio.status === H.NA ? 'No requests' : reqErrorsRatio.value.toFixed(2) + '%';
    return (
      <Popover id={this.props.id + '-health-tooltip'} title={this.props.headline}>
        <strong>
          {this.renderStatus(globalDeplStatus)}
          {' Deployments Status:'}
        </strong>
        <ul style={{ listStyleType: 'none', paddingLeft: 12 }}>
          {health.deploymentStatuses.map((st, idx) => {
            return (
              <li key={idx}>
                {this.renderStatus(deplsStatuses[idx])} {st.name}: {st.available} / {st.replicas}
              </li>
            );
          })}
        </ul>
        <strong>
          {this.renderStatus(H.mergeStatus(envoyInbound, envoyOutbound))}
          {' Envoy Health:'}
        </strong>
        <ul style={{ listStyleType: 'none', paddingLeft: 12 }}>
          <li key="inbound">
            {this.renderStatus(envoyInbound)} Inbound: {health.envoy.inbound.healthy} / {health.envoy.inbound.total}
          </li>
          <li key="outbound">
            {this.renderStatus(envoyOutbound)} Outbound: {health.envoy.outbound.healthy} / {health.envoy.outbound.total}
          </li>
        </ul>
        <strong>
          {this.renderStatus(reqErrorsRatio.status)}
          {' Error Rate:'}
        </strong>
        {' ' + reqErrorsText + ' over ' + getName(this.props.rateInterval).toLowerCase()}
      </Popover>
    );
  }

  renderStatus(status: H.Status) {
    if (status.icon) {
      return <Icon type="pf" name={status.icon} />;
    } else {
      return <span style={{ color: status.color }}>{status.text}</span>;
    }
  }
}
