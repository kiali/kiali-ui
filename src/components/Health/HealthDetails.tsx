import * as React from 'react';
import * as H from '../../types/Health';
import { createIcon } from './Helper';
import { InfoAltIcon } from '@patternfly/react-icons';
import './Health.css';
import { PfColors } from '../Pf/PfColors';

interface Props {
  health: H.Health;
  tooltip?: boolean;
}

export class HealthDetails extends React.PureComponent<Props, {}> {
  renderErrorRate = (item: H.HealthItem, idx: number) => {
    const config = this.props.health.getStatusConfig();
    const isValueInConfig =
      config && this.props.health.health.statusConfig ? this.props.health.health.statusConfig.value > 0 : false;

    const showTraffic = item.children
      ? item.children.filter(sub => {
          const showItem = sub.value && sub.value > 0;
          return (sub.status !== H.HEALTHY && showItem) || !this.props.tooltip;
        }).length > 0
      : false;
    return showTraffic ? (
      <div key={idx}>
        <strong>
          {' ' + item.title + (item.text && item.text.length > 0 ? ': ' : '')}{' '}
          {config && <InfoAltIcon color={PfColors.Gray} />}
        </strong>
        {item.text}
        {item.children && (
          <ul style={{ listStyleType: 'none', paddingLeft: 12 }}>
            {item.children.map((sub, subIdx) => {
              const showItem = sub.value && sub.value > 0;
              return (sub.status !== H.HEALTHY && showItem) || !this.props.tooltip ? (
                <li key={subIdx}>
                  {createIcon(sub.status)} {sub.text}
                </li>
              ) : (
                <></>
              );
            })}
            {config && isValueInConfig && (
              <li key={'degraded_failure_config'}>
                {createIcon(H.DEGRADED)}: {config.degraded === 0 ? '>' : '>='}
                {config.degraded}% {createIcon(H.FAILURE)}: {config.degraded === 0 ? '>' : '>='}
                {config.failure}%
              </li>
            )}
          </ul>
        )}
      </div>
    ) : (
      <></>
    );
  };

  renderChildren = (item: H.HealthItem, idx: number) => {
    return item.title.startsWith(H.TRAFFICSTATUS)
      ? this.renderErrorRate(item, idx)
      : (item.status !== H.HEALTHY || !this.props.tooltip) && (
          <div key={idx}>
            <strong>{' ' + item.title + (item.text && item.text.length > 0 ? ': ' : '')}</strong>
            {item.text}
            {item.children && (
              <ul style={{ listStyleType: 'none', paddingLeft: 12 }}>
                {item.children.map((sub, subIdx) => {
                  return (
                    <li key={subIdx}>
                      {createIcon(sub.status)} {sub.text}
                    </li>
                  );
                })}
              </ul>
            )}
          </div>
        );
  };

  render() {
    const health = this.props.health;
    return health.health.items.map((item, idx) => {
      return this.renderChildren(item, idx);
    });
  }
}
