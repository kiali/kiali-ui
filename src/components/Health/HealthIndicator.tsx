import * as React from 'react';
import { PopoverPosition, Text, TextContent, TextVariants, Tooltip } from '@patternfly/react-core';
import { HealthDetails } from './HealthDetails';
import * as H from '../../types/Health';
import { createIcon } from './Helper';
import './Health.css';
import { PfColors } from '../Pf/PfColors';

export enum DisplayMode {
  LARGE,
  SMALL
}

interface Props {
  id: string;
  health?: H.Health;
  mode: DisplayMode;
  tooltipPlacement?: PopoverPosition;
}

interface HealthState {
  globalStatus: H.Status;
}

export class HealthIndicator extends React.PureComponent<Props, HealthState> {
  static getDerivedStateFromProps(props: Props) {
    return {
      globalStatus: props.health ? props.health.getGlobalStatus() : H.NA
    };
  }

  constructor(props: Props) {
    super(props);
    this.state = HealthIndicator.getDerivedStateFromProps(props);
  }

  render() {
    if (this.props.health) {
      if (this.props.mode === DisplayMode.SMALL) {
        return this.renderSmall(this.props.health);
      } else {
        return this.renderLarge(this.props.health);
      }
    }
    return <span />;
  }

  renderSmall(health: H.Health) {
    return this.renderPopover(health, createIcon(this.state.globalStatus, 'sm'));
  }

  renderLarge(health: H.Health) {
    const spanStyle: React.CSSProperties = {
      color: this.state.globalStatus.color,
      fontWeight: 'bold',
      position: 'relative',
      top: -9,
      left: 10
    };
    return (
      <>
        {createIcon(this.state.globalStatus, 'lg')}
        <span style={spanStyle}>{this.state.globalStatus.name}</span>
        <br />
        <br />
        <HealthDetails health={health} />
      </>
    );
  }
  renderHealthTooltip(health: H.Health) {
    return (
      <TextContent style={{ color: PfColors.White }}>
        <Text component={TextVariants.h2}>{this.state.globalStatus.name}</Text>
        <HealthDetails health={health} />
      </TextContent>
    );
  }

  renderPopover(health: H.Health, icon: JSX.Element) {
    return (
      <Tooltip
        aria-label={'Health indicator'}
        content={this.renderHealthTooltip(health)}
        position={PopoverPosition.auto}
        className={'health_indicator'}
      >
        {icon}
      </Tooltip>
    );
  }
}
