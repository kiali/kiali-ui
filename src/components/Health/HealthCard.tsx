import * as React from 'react';
import { Card, CardActions, CardBody, CardHead, CardHeader, Stack, StackItem, Title } from '@patternfly/react-core';
import { DisplayMode, HealthIndicator } from './HealthIndicator';
import { style } from 'typestyle';
import * as H from '../../types/Health';

type Props = {
  name: string;
  health?: H.Health;
};

const titleStyle = style({
  margin: '15px 0 11px 0'
});

class HealthCard extends React.PureComponent<Props> {
  render() {
    return (
      <Card>
        <CardHead>
          <CardActions />
          <CardHeader>
            <Title style={{ float: 'left' }} headingLevel="h3" size="2xl">
              Health
            </Title>
          </CardHeader>
        </CardHead>
        <CardBody>
          <Stack className={'stack_service_details'}>
            <StackItem id={'health'}>
              <Title headingLevel="h3" size="lg" className={titleStyle}>
                Overall Health
              </Title>
              <HealthIndicator id={this.props.name} health={this.props.health} mode={DisplayMode.LARGE} />
            </StackItem>
          </Stack>
        </CardBody>
      </Card>
    );
  }
}

export default HealthCard;
