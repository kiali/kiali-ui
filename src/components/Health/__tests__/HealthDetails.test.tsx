import * as React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { HealthDetails } from '../HealthDetails';
import { ServiceHealth } from '../../../types/Health';

describe('HealthDetails', () => {
  it('renders healthy', () => {
    const health = new ServiceHealth(
      { errorRatio: -1, inboundErrorRatio: -1, outboundErrorRatio: -1 },
      { rateInterval: 60, hasSidecar: true }
    );

    const wrapper = shallow(<HealthDetails health={health} />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('renders deployments failure', () => {
    const health = new ServiceHealth(
      { errorRatio: -1, inboundErrorRatio: -1, outboundErrorRatio: -1 },
      { rateInterval: 60, hasSidecar: true }
    );

    const wrapper = shallow(<HealthDetails health={health} />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
