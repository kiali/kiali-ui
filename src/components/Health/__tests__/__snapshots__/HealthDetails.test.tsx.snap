// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HealthDetails renders deployments failure 1`] = `
<div
  key="0"
>
  <strong>
    <UnknownIcon
      className="icon-na"
      color="#72767b"
      noVerticalAlign={false}
      size="sm"
      title={null}
    />
     Error Rate over last 1m: 
  </strong>
  No requests
</div>
`;

exports[`HealthDetails renders healthy 1`] = `
<div
  key="0"
>
  <strong>
    <UnknownIcon
      className="icon-na"
      color="#72767b"
      noVerticalAlign={false}
      size="sm"
      title={null}
    />
     Error Rate over last 1m: 
  </strong>
  No requests
</div>
`;
